walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,144 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
 
 public final class FindMeetingQuery {
+  public static final int TOTAL_SLOTS = 24 * 60;
+  public static final int DUMMY_DURATION = 1;
+
+  public Collection<TimeRange> query1(Collection<Event> events, MeetingRequest request) {
+    long duration = request.getDuration();
+    Collection<TimeRange> answer = new ArrayList<>();
+
+    if (duration > TOTAL_SLOTS) {
+      return answer;
+    }
+
+    // build in O(PlogP) so contains() run in O(logP) instead of O(P), where P is the max number of Attendees per Event
+    TreeSet<String> attendees = new TreeSet(request.getAttendees());
+
+    // Represents whether each slot of 1 minute is occupied or not
+    boolean[] isOccupied = new boolean[TOTAL_SLOTS];  // O(1)
+
+    for (Event e : events) { // O(E), where E is the max number of Events supplied to the query
+        TimeRange timeRange = e.getWhen();
+        Set<String> participants = e.getAttendees();
+        for (String p : participants) { // O(P)
+            if (attendees.contains(p)) { // O(logP)
+                for (int i = timeRange.start(); i < timeRange.end(); i++) { // O(1)
```

Comment:
Hmm, given a for loop O(1), I guess you were saying the max number of slots is fixed.

Reply:
Yes, if we say the max slots if fixed (number of minutes in a day), then this is the fastest time complexity I think.

But, of course, the number of minutes a day is very large, likely >> E and P, or if we want to tweak such that we use continuous time slots, then the 2nd algorithm will prevail, though more complicated